File: Two pointer/15. 三数之和/15. 三数之和.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:return []
        nums.sort()
        res = []
        for i in range(len(nums)-2):
            if nums[i] > 0:break
            if i > 0 and (nums[i - 1] == nums[i]):
                continue
            l = i + 1
            r = len(nums) - 1
            while l < r:
                if nums[i]+nums[l]+nums[r] == 0:
                    if [nums[i], nums[l], nums[r]] not in res:
                        res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l+1]:
                        l += 1
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1                    
                    l += 1
                    r -= 1
                elif nums[i]+nums[l]+nums[r] > 0:
                    r -= 1
                else:
                    l += 1
        return res

File: 2. Two pointer/15. 三数之和/15. 三数之和.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> res;
        if(n<3){
            return res;
        }
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n-2; i++){
            if(nums[i]>0){
                break;
            }
            // 確保i不重複
            if(i>0 && nums[i-1] == nums[i]){
                continue;
            }
            int l = i+1;
            int r = n-1;
            while(l < r){
                int sum = nums[i]+nums[l]+nums[r];
                if(sum == 0){
                    res.push_back({nums[i],nums[l],nums[r]});
                    //確保無重複解
                    while(l < r && nums[l]==nums[l+1]){
                        l++;
                    }
                    while(l < r && nums[r]==nums[r-1]){
                        r--;
                    }
                    l++;
                    r--;
                }
                else if(sum < 0){
                    l++;
                }
                else{
                    r--;
                }
            }
        }
    return res;
    }
};